mockwebserver/src/test/java/mockwebserver3/rawsocketclient/RawSocketClient.java

```diff
@@ -0,0 +1,42 @@
+package mockwebserver3.rawsocketclient;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.Socket;
+
+import static java.net.InetAddress.getLoopbackAddress;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class RawSocketClient implements AutoCloseable {
```

Comment:
Not sure about RawSocketClient as a top level class.  If it's required for one specific test then it should probably be in that test.  If it's intended for reuse then I'm concerned about the maintenance of the tests, ideally we could operate at a higher level above a raw socket.

I don't have a clear answer, just giving my initial thoughts.

Reply:
I would like to use it in other tests. It would probably require some hairy reflection to wrap socket used by other clients by sniffer.

I think it is important to be able to make assertions during exchanges, for example to make sure that server responds after receiving request. It would require to make assertions on whole conversation, something like automatic tcpdump during tests. That probably would be hard and not so readable.

I think it should be used only:
- to test MockWebServerTest during improper client behavior so we can make sure that it is recognized and it will be visible if OkHttp behaves this way
- to test exchange with multiple requests/responses with are hidden by clients (like with 1xx codes)